From 86fc0507317d4558cc952f8e6aca3eb9f144e8b3 Mon Sep 17 00:00:00 2001
From: Guochun Huang <hero.huang@rock-chips.com>
Date: Tue, 16 Jan 2024 07:32:21 +0000
Subject: [PATCH 07/11] misc: rk628: fix display anormal after resume form
 suspend

rk628->fb_nb.notifier_call helper will be called multi times in
Multi-display suspend/resume scenario, process fb_event with
FB_BLANK_UNBLANK/FB_BLANK_POWERDOWN only once.

Type: Fix
Redmine ID: #462145
Associated modifications: gerrit links
Test: pass

Change-Id: Id323a392b28892c8440c813571a7808bb6e860ed
Signed-off-by: Guochun Huang <hero.huang@rock-chips.com>
---
 drivers/misc/rk628/rk628.c | 4 ++++
 drivers/misc/rk628/rk628.h | 1 +
 2 files changed, 5 insertions(+)

diff --git a/drivers/misc/rk628/rk628.c b/drivers/misc/rk628/rk628.c
index 4c00b25682c077..8400b3793a4606 100644
--- a/drivers/misc/rk628/rk628.c
+++ b/drivers/misc/rk628/rk628.c
@@ -533,6 +533,10 @@ static int rk628_fb_notifier_callback(struct notifier_block *nb,
 		return NOTIFY_DONE;
 
 	blank = ev_data->data;
+	if (rk628->old_blank == *blank)
+		return NOTIFY_DONE;
+
+	rk628->old_blank = *blank;
 
 	switch (*blank) {
 	case FB_BLANK_UNBLANK:
diff --git a/drivers/misc/rk628/rk628.h b/drivers/misc/rk628/rk628.h
index e5781b999dc203..973b431760fa36 100644
--- a/drivers/misc/rk628/rk628.h
+++ b/drivers/misc/rk628/rk628.h
@@ -583,6 +583,7 @@ struct rk628 {
 	struct notifier_block fb_nb;
 	u32 version;
 	struct regulator *vccio_rgb;
+	int old_blank;
 };
 
 static inline bool rk628_input_is_hdmi(struct rk628 *rk628)
-- 
2.17.1

